File: programmers/Level2/타겟넘버.java
class Solution {
    int count = 0;
    
    public int solution(int[] numbers, int target) {
        dfs(numbers, target, 0);
        return count;
    }
    
    public void dfs(int[] numbers, int target, int depth) {
        if(depth == numbers.length) {
            int sum = 0;
            for(int i : numbers) {
                sum += i;
            }
            if(sum == target) count++; 
            return;
        }
        
        numbers[depth] *= 1;
        dfs(numbers, target, depth + 1);
        numbers[depth] *= -1;
        dfs(numbers, target, depth + 1);
    }
}

File: programmers/Level2/타겟넘버.py
# # deque를 이용한 BFS 풀이
# from collections import deque
# def solution(numbers, target):
#     answer = 0
#     n = len(numbers)
#     queue = deque()
#     queue.append([numbers[0], 0])         # value, idx
#     queue.append([-1 * numbers[0], 0])    # value, idx
#     while queue:
#         temp, idx = queue.popleft()
#         idx += 1
#         if idx < n:
#             queue.append([temp + numbers[idx], idx])
#             queue.append([temp - numbers[idx], idx])
#         else:
#             if temp == target:
#                 answer += 1
#     return answer

# stack을 이용한 dfs 풀이
def solution(numbers, target):
    answer = 0
    n = len(numbers)
    stack = [[numbers[0], 0], [-1 * numbers[0], 0]]
    
    while stack:
        temp, idx = stack.pop()
        idx += 1
        if idx < n:
            stack.append([temp + numbers[idx], idx])
            stack.append([temp - numbers[idx], idx])
        else:
            if temp == target:
                answer += 1
                
    return answer